crates/ruma-client-api/src/room/get_summary.rs

```diff
@@ -0,0 +1,150 @@
+        history: {
+            unstable => "/_matrix/client/unstable/im.nheko.summary/rooms/:room_id_or_alias/summary",
+        }
+    };
+
+    /// Request type for the `summary/room_id_or_alias` endpoint.
+    #[request(error = crate::Error)]
+    pub struct Request {
+        /// Alias or ID of the room to be summarized.
+        #[ruma_api(path)]
+        pub room_id_or_alias: OwnedRoomOrAliasId,
+
+        /// An optional list of servers the invited homeserver should attempt to peek at the room.
+        ///
+        /// If present, it must not be empty.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub via: Option<Vec<OwnedServerName>>,
+    }
+
+    /// Response type for the `summary/room_id_or_alias` endpoint.
+    #[response(error = crate::Error)]
+    pub struct Response {
+        /// ID of the room (useful if it's an alias).
+        pub room_id: OwnedRoomId,
+
+        /// A canonical alias for this room, if set.
+        pub canonical_alias: Option<OwnedRoomAliasId>,
+
+        /// Avatar of the room.
+        pub avatar_url: Option<OwnedMxcUri>,
+
+        /// Whether guests can join the room.
+        pub guest_can_join: bool,
+
+        /// Name of the room.
+        pub name: Option<String>,
+
+        /// Member count of the room.
+        pub num_joined_members: u64,
+
+        /// Topic of the room.
+        pub topic: Option<String>,
+
+        /// Whether the room history can be read without joining.
+        pub world_readable: bool,
+
+        /// Join rule of the room.
+        #[serde(deserialize_with = "deserialize_as_join_rule")]
+        pub join_rule: JoinRule,
+
+        /// Type of the room, if any.
+        pub room_type: Option<RoomType>,
+
+        /// Version of the room.
+        pub room_version: Option<RoomVersionId>,
+
+        /// The current membership of this user in the room. Usually `leave` if the room is fetched
+        /// over federation.
+        ///
+        /// This field will not be present when called unauthenticated, but is required when called
+        /// authenticated. It should be `leave` if the serer doesn't know about the room, since for
+        /// all other membership states the server would know about the room already.
+        pub membership: Option<MembershipState>,
+
+        /// If the room is encrypted, this specificed the algorithm used for this room.
+        pub encryption: Option<EventEncryptionAlgorithm>,
+    }
+
+    impl Request {
+        /// Creates a new `Request` with the given room or alias ID and via server names.
+        pub fn new(
+            room_id_or_alias: OwnedRoomOrAliasId,
+            via: Option<Vec<OwnedServerName>>,
+        ) -> Self {
+            Self { room_id_or_alias, via }
+        }
+    }
+
+    #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+    #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+    #[serde(rename_all = "snake_case")]
+    enum FlatJoinRule {
```

Comment:
Why this special type?

Reply:
it was required before `JoinRule` requires jumping through many hops to be correctly deserialized (`#[serde(flatten)]` + custom `Deserialize` impl for the whole structure containing it), and we _think_ from parsing the MSC that the `allow` rules wouldn't be contained in the summary response

Reply:
As discussed on Matrix, this doesn't seem to be actually necessary, this new endpoint can use `SpaceRoomJoinRule` (can be renamed / moved in this PR if you want).